File: algorithms/031.Next_Permutation/fractional_1.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    void nextPermutation(vector<int> &num) {
        if (!num.size())
            return;

        int idx = num.size() - 2;

        // 1. find out the last wrong order
        while (idx >= 0 && num[idx] >= num[idx + 1])
            idx--;

        // 2. swap
        if (idx >= 0) {
            int i = idx + 1;
            while (i < num.size() && num[i] > num[idx])
                i++;
            swap(num[i - 1], num[idx]);
        }

        // 3. reverse
        reverse(num.begin() + idx + 1, num.end());
    }
};

int main() {
    int n;
    Solution s;
    while (cin >> n) {
        vector<int> num(n);
        for (auto &i : num)
            cin >> i;
        s.nextPermutation(num);
        for (auto &i : num)
            cout << i << ' ';
        cout << endl;
    }
    return 0;
}

File: algorithms/031.Next_Permutation/fractional_1.py
# 一个全排列问题,主要是先找最后一个逆序数，然后反转
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        k = -1
        for i in range(len(nums)-2, -1, -1):
            if nums[i] < nums[i+1]:
                k = i
                break
        print(k)
        if k == -1:
            nums.reverse()
            return
        l = -1
        for i in range(len(nums)-1, k, -1):
            if nums[i] > nums[k]:
                l = i
                break
        print(l)
        nums[l], nums[k] = nums[k], nums[l]
        self.reverse(nums,k+1,len(nums)-1)
    def reverse(self,nums,l,r):
        while l < r:
            nums[l],nums[r] = nums[r],nums[l]
            l += 1
            r -= 1

a=[1,3,2]
s= Solution()
s.nextPermutation(a)
print(a)